### PR TITLE
[Pulsar Functions] Add support for multiple sinks/sources in a single NAR file.

### PR DESCRIPTION
**Pull request by david-streamlio**
### Contribution Checklist

Address https://github.com/apache/pulsar/issues/3576

### Motivation

Added logic to use the proper class loader when a Source or Sink is created from the command line and an "archive" and "className" are provided. Prior to this change, it was always assumed the the archive file specified was a standard "JAR" file. 

Now, we check the file extension of the archive and choose the appropriate class loader based on that. This allows user to submit Pulsar connectors that are bundled inside a NAR file

Also added support for multiple sink/source definitions in NAR files, while also being 100% fully-backward complaint. Added test classes to validate that

### Modifications

Modified both the 
pulsar-functions/utils/src/main/java/org/apache/pulsar/functions/utils/SinkConfigUtils.java
pulsar-functions/utils/src/main/java/org/apache/pulsar/functions/utils/SourceConfigUtils.java

classes to check the file extension of the archive, and use the correct class loader

Modified the org.apache.pulsar.functions.utils.io.ConnectorUtils class to read both the old and new pulsar-io.yaml formats as shown below. Therefore this is NOT a breaking change

`name: file`
`description: Reads data from local filesystem`
`sourceClass: org.apache.pulsar.io.file.FileSource`

and 

`connectors:`
`                hdfs:`
`                 description: The HDFS connector `
`                  sourceClass: org.apache.pulsar.io.hdfs.SoruceClassImpl`
`                  sinkClass: org.apache.pulsar.io.hdfs.SinkClassImpl `
`                redis:`
`                  description: The Redis connector`
`                  sourceClass: org.apache.pulsar.io.redis.SoruceClassImpl `
`                  sinkClass: org.apache.pulsar.io.redis.SinkClassImpl`

### Verifying this change

- [ ] Make sure that the change passes the CI checks.

This change is already covered by existing tests, such as

- org.apache.pulsar.functions.worker.rest.api.v3.SinkApiV3ResourceTest
- org.apache.pulsar.functions.worker.rest.api.v3.SourceApiV3ResourceTest

### Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): No
  - The public API: No
  - The schema: No
  - The default values of configurations: No
  - The wire protocol: No
  - The rest endpoints: No (Fixes existing broken usage of archive/className params)
  - The admin cli options: No (Fixes existing broken usage of archive/className params)
  - Anything that affects deployment: No

### Documentation

  - Does this pull request introduce a new feature? No
  - If yes, how is the feature documented? not applicable
  - If a feature is not applicable for documentation, explain why? It's already documented

### CODE REVIEW

File: conf/pulsar_tools_env.sh [hunk 1]
```diff
@@ -42,7 +42,8 @@
 # PULSAR_GLOBAL_ZK_CONF=
 
 # Extra options to be passed to the jvm
-PULSAR_MEM="-Xmx128m -XX:MaxDirectMemorySize=128m"
+# PULSAR_MEM="-Xmx128m -XX:MaxDirectMemorySize=128m"
+PULSAR_MEM="-Xmx4g -XX:MaxDirectMemorySize=128m"
```
Comment (jerrypeng) [inline conf/pulsar_tools_env.sh]: is this change necessary?

**Comment by jerrypeng** [comment]
@david-streamlio so which multiple sources/sinks packaged in a single NAR I would use the specify the classname to run a particular one? e.g.

```
./bin/pulsar-admin source create --name test --destinationTopicName output --archive foo.nar --classname my.source

```

File: conf/pulsar_tools_env.sh (hunk 1 above)
Comment (david-streamlio) [reply conf/pulsar_tools_env.sh]: No, I had to add that in order to be able to submit the NAR I was testing with.

**Comment by david-streamlio** [comment]
Most connectors require multiple supporting jars to be bundled with them. This allows you to submit either an Uber jar or a NAR file.

I used the pulsar-aws-nar-0.0.1.nar file for testing, which has 3 sink and 3 source connectors.

**Comment by jerrypeng** [comment]
@david-streamlio this PR looks good to me, but I am unclear how the changes in the PR makes submitting a uber JAR with multiple sources/sinks different from submitting a NAR.  Do we need to support submitting with a NAR with multiple sources/sinks if submitting a uber JAR already covers the use cases of multiple sources/sinks in one package? I guess it doesn't hurt, however my concern is that we have always submitted NARs without specifying a classname since there is the a yaml file under META-INF that specifies the default source/sink classname as well as the name of the source/sink.  How does that work here?  with multiple sources and sinks.

**Comment by david-streamlio** [comment]
There wouldn't be a difference between submitting an uber jar and a NAR. They would both be submitted in the exact same way. Each would require the user to specify an archive AND a classname. 

`bin/pulsar-admin source create --classname org.apache.pulsar.io.aws.sqs.SqsSource --archive /Users/david/clone-zone/streamlio/pulsar-connectors/pulsar-nar-bundles/pulsar-aws-connector-bundle/pulsar-aws-nar/target/pulsar-aws-nar-0.0.1.nar --className org.apache.pulsar.io.aws.s3.source.AwsS3Source --name s3-source --destination-topic-name s3_data --source-config-file /Users/david/Desktop/s3-config.yaml`

**Comment by jerrypeng** [comment]
@david-streamlio so for this NAR you built, pulsar-aws-nar-0.0.1.nar,  with multiple sources and sinks, what happens if I submit this NAR without a classname? what happens? I ask since currently submitting a NAR without specifying the classname is supported and the de factor way to submit a NAR.  Does it contain a yaml file under META-INF that specifies the default source/sink classname like the rest of the sources and sinks that we currently have?

**Comment by sijie** [comment]
@jerrypeng @david-streamlio - did you have any conclusions in #3576? It would be worth that you guys can conclude an approach in #3576 , so that everyone is aligned with the approach and other committers can also review it based on that approach.

**Comment by david-streamlio** [comment]
@jerrypeng If you submit the pulsar-aws-nar-0.0.1.nar without specifying a classname, then the sink/source configured in the pulsar-io.yaml file will be used. You can think of that as the "default" connectors, that can be overridden by the className property.

**Comment by jerrypeng** [comment]
@david-streamlio ok that sounds good to me. Just making sure all code paths are covered.

**Comment by jerrypeng** [comment]
@david-streamlio please revert back the change in pulsar_tools_env.sh and I think we are good to go. Thanks!

**Review by jerrypeng — DISMISSED** [review]

**Comment by jiazhai** [comment]
rerun java8 tests

File: pulsar-functions/utils/src/main/java/org/apache/pulsar/functions/utils/SinkConfigUtils.java [hunk 2]
```diff
@@ -297,7 +299,21 @@ public static ExtractedSinkDetails validate(SinkConfig sinkConfig, Path archiveP
         if (!isEmpty(sinkConfig.getClassName())) {
             sinkClassName = sinkConfig.getClassName();
             try {
-                classLoader = Utils.extractClassLoader(archivePath, functionPkgUrl, uploadedInputStreamAsFile);
+                String ext = null;
+            	if (!isEmpty(functionPkgUrl)) {
+            		File file = Utils.extractFileFromPkg(functionPkgUrl);
```
Comment (sijie) [inline pulsar-functions/utils/src/main/java/org/apache/pulsar/functions/utils/SinkConfigUtils.java]: spaces instead of tabs?

File: pulsar-functions/utils/src/main/java/org/apache/pulsar/functions/utils/SinkConfigUtils.java [hunk 3]
```diff
@@ -297,7 +299,21 @@ public static ExtractedSinkDetails validate(SinkConfig sinkConfig, Path archiveP
...
+            		ext = FilenameUtils.getExtension(file.getName());
```
Comment (sijie) [inline pulsar-functions/utils/src/main/java/org/apache/pulsar/functions/utils/SinkConfigUtils.java]: spaces instead of tabs?

File: pulsar-functions/utils/src/main/java/org/apache/pulsar/functions/utils/SinkConfigUtils.java [hunk 4]
```diff
@@ -297,7 +299,21 @@ public static ExtractedSinkDetails validate(SinkConfig sinkConfig, Path archiveP
...
+            	} else if (archivePath != null) {
+            		ext = FilenameUtils.getExtension(archivePath.getFileName().toString());
```
Comment (sijie) [inline pulsar-functions/utils/src/main/java/org/apache/pulsar/functions/utils/SinkConfigUtils.java]: same as above

File: pulsar-functions/utils/src/main/java/org/apache/pulsar/functions/utils/SinkConfigUtils.java [hunk 5]
```diff
@@ -297,7 +299,21 @@ public static ExtractedSinkDetails validate(SinkConfig sinkConfig, Path archiveP
...
+            	} else if (uploadedInputStreamAsFile != null) {
+            		ext = FilenameUtils.getExtension(uploadedInputStreamAsFile.getName());
```
Comment (sijie) [inline pulsar-functions/utils/src/main/java/org/apache/pulsar/functions/utils/SinkConfigUtils.java]: same as above

File: pulsar-functions/utils/src/main/java/org/apache/pulsar/functions/utils/SinkConfigUtils.java [hunk 6]
```diff
@@ -297,7 +299,21 @@ public static ExtractedSinkDetails validate(SinkConfig sinkConfig, Path archiveP
...
+            	}
+            	
+            	if ("nar".equalsIgnoreCase(ext)) {
+            	   classLoader = Utils.extractNarClassLoader(archivePath, functionPkgUrl, uploadedInputStreamAsFile);
```
Comment (sijie) [inline pulsar-functions/utils/src/main/java/org/apache/pulsar/functions/utils/SinkConfigUtils.java]: wrong indent

File: pulsar-functions/utils/src/main/java/org/apache/pulsar/functions/utils/SourceConfigUtils.java [hunk 7]
```diff
@@ -210,7 +212,21 @@ public static ExtractedSourceDetails validate(SourceConfig sourceConfig, Path ar
         if (!isEmpty(sourceConfig.getClassName())) {
             sourceClassName = sourceConfig.getClassName();
             try {
-                classLoader = Utils.extractClassLoader(archivePath, functionPkgUrl, uploadedInputStreamAsFile);
+            	String ext = null;
+            	if (!isEmpty(functionPkgUrl)) {
+            		File file = Utils.extractFileFromPkg(functionPkgUrl);
```
Comment (sijie) [inline pulsar-functions/utils/src/main/java/org/apache/pulsar/functions/utils/SourceConfigUtils.java]: same comments regarding tabs as above

File: pulsar-functions/utils/src/main/java/org/apache/pulsar/functions/utils/SinkConfigUtils.java (hunk 2 above)
Comment (david-streamlio) [reply pulsar-functions/utils/src/main/java/org/apache/pulsar/functions/utils/SinkConfigUtils.java]: Fixed

File: pulsar-functions/utils/src/main/java/org/apache/pulsar/functions/utils/SinkConfigUtils.java (hunk 3 above)
Comment (david-streamlio) [reply pulsar-functions/utils/src/main/java/org/apache/pulsar/functions/utils/SinkConfigUtils.java]: Fixed

File: pulsar-functions/utils/src/main/java/org/apache/pulsar/functions/utils/SinkConfigUtils.java (hunk 4 above)
Comment (david-streamlio) [reply pulsar-functions/utils/src/main/java/org/apache/pulsar/functions/utils/SinkConfigUtils.java]: Fixed

File: pulsar-functions/utils/src/main/java/org/apache/pulsar/functions/utils/SinkConfigUtils.java (hunk 5 above)
Comment (david-streamlio) [reply pulsar-functions/utils/src/main/java/org/apache/pulsar/functions/utils/SinkConfigUtils.java]: Fixed

File: pulsar-functions/utils/src/main/java/org/apache/pulsar/functions/utils/SinkConfigUtils.java (hunk 6 above)
Comment (david-streamlio) [reply pulsar-functions/utils/src/main/java/org/apache/pulsar/functions/utils/SinkConfigUtils.java]: Resolved

File: pulsar-functions/utils/src/main/java/org/apache/pulsar/functions/utils/SourceConfigUtils.java (hunk 7 above)
Comment (david-streamlio) [reply pulsar-functions/utils/src/main/java/org/apache/pulsar/functions/utils/SourceConfigUtils.java]: Resolved

File: pulsar-functions/utils/src/main/java/org/apache/pulsar/functions/utils/SourceConfigUtils.java [hunk 8]
```diff
@@ -210,7 +212,21 @@ public static ExtractedSourceDetails validate(SourceConfig sourceConfig, Path ar
         if (!isEmpty(sourceConfig.getClassName())) {
             sourceClassName = sourceConfig.getClassName();
             try {
-                classLoader = Utils.extractClassLoader(archivePath, functionPkgUrl, uploadedInputStreamAsFile);
+                String ext = null;
+                if (!isEmpty(functionPkgUrl)) {
+                    File file = Utils.extractFileFromPkg(functionPkgUrl);
```
Comment (eolivelli) [inline pulsar-functions/utils/src/main/java/org/apache/pulsar/functions/utils/SourceConfigUtils.java:217]: this file is only a temporary file, we need it only for the extension.
can we remove the file as soon as we do not need it anymore ?
it would be also better to not extract the file but only to scan for the file name

File: pulsar-functions/worker/src/test/java/org/apache/pulsar/functions/worker/rest/api/v3/SourceApiV3ResourceTest.java [hunk 9]
```diff
@@ -105,7 +106,7 @@ public IObjectFactory getObjectFactory() {
     private static final String className = TwitterFireHose.class.getName();
     private static final int parallelism = 1;
     private static final String JAR_FILE_NAME = "pulsar-io-twitter.nar";
-    private static final String INVALID_JAR_FILE_NAME = "pulsar-io-cassandra.nar";
+    private static final String INVALID_JAR_FILE_NAME = "pulsar-io-cassandra-2.3.0-SNAPSHOT.nar";
```
Comment (eolivelli) [inline pulsar-functions/worker/src/test/java/org/apache/pulsar/functions/worker/rest/api/v3/SourceApiV3ResourceTest.java:109]: why are you changing this line ? does the version make a difference to us ?

File: pulsar-functions/worker/src/test/java/org/apache/pulsar/functions/worker/rest/api/v3/SourceApiV3ResourceTest.java [hunk 10]
```diff
@@ -931,11 +932,16 @@ public void testUpdateSourceWithUrl() throws IOException {
...
         mockStatic(ConnectorUtils.class);
         doReturn(TwitterFireHose.class.getName()).when(ConnectorUtils.class);
-        ConnectorUtils.getIOSourceClass(any(NarClassLoader.class));
+        ConnectorUtils.getIOSourceClass(anyString(), any(NarClassLoader.class));
 
         mockStatic(org.apache.pulsar.functions.utils.Utils.class);
         doReturn(String.class).when(org.apache.pulsar.functions.utils.Utils.class);
         org.apache.pulsar.functions.utils.Utils.getSourceType(anyString(), any(NarClassLoader.class));
+        
+        File mockedFile = mock(File.class);
```
Comment (eolivelli) [inline pulsar-functions/worker/src/test/java/org/apache/pulsar/functions/worker/rest/api/v3/SourceApiV3ResourceTest.java:941]: usually it is not a good practice to mock Java classes, it would be likely to break with future Java version, in this case you can create a dummy temporary file with the wanted file extension

**Review by eolivelli — CHANGES_REQUESTED** [review]
@david-streamlio this is a very nice an useful feature.
I left a few little comments, can you please take a look and address the comments from the others ?

it would be great to see this land in 2.8.0

**Comment by tisonkun** [comment]
Closed as stale.
